From 7e2a348a0c67d926b2bd63d2cf9e6b9d84ae1585 Mon Sep 17 00:00:00 2001
From: Maciej Szymanski <mszymans@qti.qualcomm.com>
Date: Mon, 15 Jun 2026 15:03:19 +0200
Subject: [PATCH] vdec: Keep last port reconfiguration event in the queue when
 flushing output port.

Keep the last port reconfiguration event received from hardware when
performing output port flushing in order to notify the client that it
needs to reconfigure the port.

Change-Id: Id997d9eb21b986d12503a93439143aef157b481c
Signed-off-by: Andrii Cherniavskyi <achernia@qti.qualcomm.com>
Signed-off-by: Maciej Szymanski <mszymans@qti.qualcomm.com>
---
 mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp b/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
index 253a60d..bf398c2 100755
--- a/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
+++ b/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
@@ -2762,7 +2762,10 @@ bool omx_vdec::execute_output_flush()
     unsigned long p1 = 0; // Parameter - 1
     unsigned long p2 = 0; // Parameter - 2
     unsigned long ident = 0;
+    unsigned long tmp_p1 = 0;
+    unsigned long tmp_p2 = 0;
     bool bRet = true;
+    bool bReconfig = false;
     unsigned int buf_index = 0;
     OMX_BUFFERHEADERTYPE *buffer = NULL;
 
@@ -2791,8 +2794,21 @@ bool omx_vdec::execute_output_flush()
             m_cb.FillBufferDone(&m_cmp, m_app_data, buffer);
         } else if (ident == OMX_COMPONENT_GENERATE_FBD) {
             fill_buffer_done(&m_cmp,(OMX_BUFFERHEADERTYPE *)(intptr_t)p1);
+        } else if (ident == OMX_COMPONENT_GENERATE_PORT_RECONFIG) {
+            bReconfig = true;
+            tmp_p1 = p1;
+            tmp_p2 = p2;
         }
     }
+
+    /* This is a workaround for a race occuring when flush is requested and port reconfig
+     * event is received from the hardware at the same time: the reconfig event will not
+     * be delivered to the client but the hardware will keep waiting for the port
+     * reconfig sequence. Hence, the event is preserved in the queue, so that it will be
+     * sent to the client at the next iteration. */
+    if (bReconfig)
+        m_ftb_q.insert_entry(tmp_p1, tmp_p2, OMX_COMPONENT_GENERATE_PORT_RECONFIG);
+
     pthread_mutex_unlock(&m_lock);
     output_flush_progress = false;
 
-- 
2.49.0

